File: 0394-decode-string/0394-decode-string.cpp
#include <string>
#include <stack>

using namespace std;

class Solution {
public:
    string decodeString(string s) {
        stack<int> counts;
        stack<string> strings;
        string result;
        int count = 0;
        
        for (char c : s) {
            if (isdigit(c)) {
                count = count * 10 + (c - '0'); // Accumulate the count
            } else if (c == '[') {
                counts.push(count);
                strings.push(result);
                result = ""; // Reset result for the new nested string
                count = 0; // Reset count
            } else if (c == ']') {
                int repeatCount = counts.top();
                counts.pop();
                string repeatedString = result;
                for (int i = 1; i < repeatCount; ++i) {
                    repeatedString += result;
                }
                result = strings.top() + repeatedString;
                strings.pop();
            } else {
                result += c; // Accumulate characters
            }
        }
        
        return result;
    }
};

File: 0394-decode-string/0394-decode-string.java
import java.util.Stack;

class Solution {
    public String decodeString(String s) {
        Stack<Integer> counts = new Stack<>();
        Stack<String> strings = new Stack<>();
        String result = "";
        int count = 0;

        for (char c : s.toCharArray()) {
            if (Character.isDigit(c)) {
                // Accumulate the count (handles multi-digit numbers)
                count = count * 10 + (c - '0');
            } else if (c == '[') {
                // Push current state to stacks and reset
                counts.push(count);
                strings.push(result);
                result = "";
                count = 0;
            } else if (c == ']') {
                // Pop repeat count and previous string segment
                int repeatCount = counts.pop();
                StringBuilder repeatedString = new StringBuilder(result);
                String temp = result;
                
                // Perform the repetition
                for (int i = 1; i < repeatCount; i++) {
                    repeatedString.append(temp);
                }
                
                // Prepend the string segment before the current brackets
                result = strings.pop() + repeatedString.toString();
            } else {
                // Accumulate characters
                result += c;
            }
        }

        return result;
    }
}